docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+
+More complex statements don't need to be supported.
+
+The type of generic classes can be narrowed by annotating the ``self``
+argument of the ``__init__`` method::
+
+    class Foo(Generic[_T]):
+        @overload
+        def __init__(self: Foo[str], type: Literal["s"]) -> None: ...
+        @overload
+        def __init__(self: Foo[int], type: Literal["i"]) -> None: ...
+        @overload
+        def __init__(self, type: str) -> None: ...
+
+The class must match the class in which it is declared. Using other classes,
+including sub or super classes, will not work. In addition, the ``self``
+annotation cannot contain type variables.
+
+Functions and Methods
+"""""""""""""""""""""
 
-* Modules that must be compatible with Python 2 and 3
+Function and method definition syntax follows general Python syntax.
+For backwards compatibility, positional-only parameters can also be marked by
+prefixing their name with two underscores (but not suffixing it with two
+underscores)::
 
-* Modules that use annotations for other purposes
+    # x is positional-only
+    # y can be used positionally or as keyword argument
+    # z is keyword-only
+    def foo(x, /, y, *, z): ...  # recommended
+    def foo(__x, y, *, z): ...  # backwards compatible syntax
 
-Stub files have the same syntax as regular Python modules.  There is one
-feature of the ``typing`` module that is different in stub files:
-the ``@overload`` decorator described below.
+If an argument or return type is unannotated, per :pep:`484` its
+type is assumed to be ``Any``. It is preferred to leave unknown
+types unannotated rather than explicitly marking them as ``Any``, as some
+type checkers can optionally warn about unannotated arguments.
 
-The type checker should only check function signatures in stub files;
-It is recommended that function bodies in stub files just be a single
-ellipsis (``...``).
+If an argument has a literal or constant default value, it must match the implementation
+and the type of the argument (if specified) must match the default value.
+Alternatively, ``...`` can be used in place of any default value::
 
-The type checker should have a configurable search path for stub files.
-If a stub file is found the type checker should not read the
-corresponding "real" module.
+    # The following arguments all have type Any.
+    def unannotated(a, b=42, c=...): ...
+    # The following arguments all have type int.
+    def annotated(a: int, b: int = 42, c: int = ...): ...
+    # The following default values are invalid and the types are unspecified.
+    def invalid(a: int = "", b: Foo = Foo()): ...
 
-While stub files are syntactically valid Python modules, they use the
-``.pyi`` extension to make it possible to maintain stub files in the
-same directory as the corresponding real module.  This also reinforces
-the notion that no runtime behavior should be expected of stub files.
+For a class ``C``, the type of the first argument to a classmethod is
```

Comment:
This doesn't belong in this section. The inferred type of `self` and `cls` should be covered in other parts of the typing spec, not in the stubs section. (It's not currently covered, but I hope to fix that in the near future.)

Regardless, I think this statement is incorrect. The type of the first argument to a classmethod is assumed to be `type[Self]` if unannotated. And the type of `self` is assumed to be `Self`.